helix-core/src/main/java/org/apache/helix/manager/zk/ZkBaseDataAccessor.java

```diff
@@ -166,14 +204,29 @@ public boolean set(String path, T record, int options) {
    */
   @Override
   public boolean set(String path, T record, int expectVersion, int options) {
-    AccessResult result = doSet(path, record, expectVersion, options);
+    AccessResult result = doSet(_zkClient, path, record, expectVersion, options);
+    return result._retCode == RetCode.OK;
+  }
+
+  /**
+   * Sync set operation with custom serializer support
+   */
+  public boolean set(String path, Object record, int options, ZkSerializer zkSerializer) {
+    return set(path, record, options, -1, zkSerializer);
+  }
+
+  /**
+   * Sync set operation with custom serializer support
+   */
+  public boolean set(String path, Object record, int options, int expectVersion, ZkSerializer zkSerializer) {
+    AccessResult result = doSet(getNonZNRecordZkClient(), path, zkSerializer.serialize(record), expectVersion, options);
     return result._retCode == RetCode.OK;
   }
 
   /**
    * sync set
    */
-  public AccessResult doSet(String path, T record, int expectVersion, int options) {
```

Comment:
We need to keep the old public API method signature. Otherwise, this diff will not be backward-compatible. Could you refactor so that we still have all the old public API methods that has the same behavior?

See my comment on doCreate. If we do this, we won't have to change the code in other places like ZkHelixCacheDataAccessor.